### PR TITLE
Add factory methods for module export/open info that is based on descriptors rather than entries

### PR DESCRIPTION
**Pull request by raphw**
Allow to lookup opcode by bytecode to ease interaction with other byte code processors that provide bytecode values.

### CODE REVIEW

File: src/java.base/share/classes/jdk/classfile/attribute/ModuleOpenInfo.java [hunk 1]
```diff
@@ -116,4 +120,56 @@ static ModuleOpenInfo of(PackageEntry opens,
...
+     * @param opens the package to open
+     * @param opensFlags the open flags
+     * @param opensTo the packages to which this package is opened, if it is a qualified open
+     */
+    static ModuleOpenInfo of(PackageDesc opens,
+                             int opensFlags,
+                             List<ModuleDesc> opensTo) {
+        return new UnboundAttribute.UnboundModuleOpenInfo(TemporaryConstantPool.INSTANCE.packageEntry(TemporaryConstantPool.INSTANCE.utf8Entry(opens.packageName())),
+                opensFlags,
+                opensTo.stream().map(o -> TemporaryConstantPool.INSTANCE.moduleEntry(TemporaryConstantPool.INSTANCE.utf8Entry(o.moduleName()))).collect(Collectors.toList()));
```
Comment (asotona) [inline src/java.base/share/classes/jdk/classfile/attribute/ModuleOpenInfo.java:135]: The same case as in ModuleExportInfo:
```
return new UnboundAttribute.UnboundModuleOpenInfo(TemporaryConstantPool.INSTANCE.packageEntry(opens),
                 opensFlags,
                 opensTo.stream().map(TemporaryConstantPool.INSTANCE::moduleEntry).toList());
```

File: src/java.base/share/classes/jdk/classfile/attribute/ModuleExportInfo.java [hunk 2]
```diff
@@ -121,4 +125,56 @@ static ModuleExportInfo of(PackageEntry exports,
...
+     * @param exports the exported package
+     * @param exportFlags the export flags
+     * @param exportsTo the modules to which this package is exported
+     */
+    static ModuleExportInfo of(PackageDesc exports,
+                               int exportFlags,
+                               List<ModuleDesc> exportsTo) {
+        return new UnboundAttribute.UnboundModuleExportInfo(TemporaryConstantPool.INSTANCE.packageEntry(TemporaryConstantPool.INSTANCE.utf8Entry(exports.packageName())),
+                exportFlags,
+                exportsTo.stream().map(e -> TemporaryConstantPool.INSTANCE.moduleEntry(TemporaryConstantPool.INSTANCE.utf8Entry(e.moduleName()))).collect(Collectors.toList()));
```
Comment (asotona) [inline src/java.base/share/classes/jdk/classfile/attribute/ModuleExportInfo.java:140]: Direct factory methods ::moduleEntry(ModuleDesc) and ::packageEntry(PackageDesc) are available and direct collector method Stream::toList():
```
return new UnboundAttribute.UnboundModuleExportInfo(TemporaryConstantPool.INSTANCE.packageEntry(exports),
                 exportFlags,
                 exportsTo.stream().map(TemporaryConstantPool.INSTANCE::moduleEntry).toList());
```

**Review by asotona — COMMENTED** [review]
### First part:
Proposed static method Opcode::ofBytecode(int) with statically initialized back array of just selected sub-set of Opcodes is non-systematic addon to the API.
Opcode enum is a primary and complete primitive describing the opcodes and suppose to act as the source of the information. 
Can you, please, describe more specifically the use case for translation of the selected opcodes from int values?

### Second part of the pull request:
Addition of missing factories to ModuleExportInfo and ModuleOpenInfo is reasonable (with just minor technical comments inline).
Questionable is if we really need all 8 factory method parameters combinations, I would recommend to cut it to the most frequently used to keep the API slim.  


**Technical note:** to refresh closed pull refresh from the original branch it is necessary to re-open it (it is annoying, however working until the auto-closing of PRs in jdk-sandbox is resolved)

**Comment by raphw** [comment]
As for (1), I want to bridge from one byte code processor to another. The first processor emits bytecodes for any instruction and adds them to a CodeBuilder. To avoid mapping each bytecode manually, I wanted to map groups of codes where I need to lookup the Opcode from a given bytecode. I did not include the pseudocodes as they are not in the spec. I do however understand that this might be too narrow for a public API and moved it to my code base.

As for (2), I made the changes. I would however suggest to keep the matrix of all factories. I find it impossible to say what method would be most commonly used, so it feels like an artificial constraint to me.

**Comment by raphw** [comment]
I don't have privilege to reopen pull requests, unfortunately.

**Comment by asotona** [comment]
My attempt to re-open this PR fails with message `The classfile-api-moduleinfo branch was force-pushed or recreated.`
The only option seems to be create a new PR.

**Pull request closed without merging**